### PR TITLE
fix: detect and fallback garbled PDF text to OCR (#13366)

### PR DESCRIPTION
**Pull request by tunsuy**
## Problem

When PDF fonts lack ToUnicode/CMap mappings, pdfplumber (pdfminer) cannot map CIDs to correct Unicode characters, outputting PUA characters (U+E000~U+F8FF) or `(cid:xxx)` placeholders. The original code fully trusted pdfplumber text without any garbled detection, causing garbled output in the final parsed result.

Relates to #13366

### 1. Garbled text detection functions
- `_is_garbled_char(ch)`: Detects PUA characters (BMP/Plane 15/16), replacement character U+FFFD, control characters, and unassigned/surrogate codepoints
- `_is_garbled_text(text, threshold)`: Calculates garbled ratio and detects `(cid:xxx)` patterns

### 2. Box-level fallback (in `__ocr()`)
When a text box has ≥50% garbled characters, discard pdfplumber text and fallback to OCR recognition.

### 3. Page-level detection (in `__images__()`)
Sample characters from each page; if garbled rate ≥30%, clear all pdfplumber characters for that page, forcing full OCR.

### 4. Layout recognizer CID filtering
Filter out `(cid:xxx)` patterns in `layout_recognizer.py` text processing to prevent them from polluting layout analysis.

## Testing
- 29 unit tests covering: normal CJK/English text, PUA characters, CID patterns, mixed text, boundary thresholds, edge cases
- All 85 existing project unit tests pass without regression

### CODE REVIEW

**Comment by Magicbook1108** [comment]
I don't think your code correctly fix the issue. Can you take a screen shot of your succeed condition? 
image
[GB.18067-2000.pdf](https://github.com/user-attachments/files/25765679/GB.18067-2000.pdf)

**Comment by tunsuy** [comment]
> I don't think your code correctly fix the issue. Can you take a screen shot of your succeed condition? image [GB.18067-2000.pdf](https://github.com/user-attachments/files/25765679/GB.18067-2000.pdf)

Updated with the fix. Here's the verification result for GB.18067-2000.pdf:

Detection: All 4 garbled pages (1, 3, 5, 8) with subset font encoding (e.g. BCDGEE+R0015) are correctly identified — CJK ratio < 5%, ASCII punctuation ratio > 40%.

Before fix: pdfplumber extracts garbled ASCII symbols (random punctuation like ()[]{} instead of Chinese text).
After fix: Garbled chars are cleared, OCR fallback is triggered, and Chinese content is successfully extracted.

Screenshot attached below. All 50 unit tests passed.

Clipboard_Screenshot_1772766111

Clipboard_Screenshot_1772766132

Clipboard_Screenshot_1772766561

Clipboard_Screenshot_1772766577

File: deepdoc/parser/pdf_parser.py [hunk 1]
```diff
@@ -1400,6 +1584,34 @@ def __images__(self, fnm, zoomin=3, page_from=0, page_to=299, callback=None):
...
 
+                    # Detect garbled pages and clear their chars so the OCR
+                    # path will be used instead. Two detection strategies:
+                    # 1) PUA / unmapped CID characters (threshold=0.3)
+                    # 2) Font-encoding garbling: subset fonts mapping CJK to ASCII
+                    for pi, page_ch in enumerate(self.page_chars):
+                        if not page_ch:
+                            continue
+                        # Strategy 1: PUA / CID garbling
+                        sample = page_ch if len(page_ch) <= 200 else random.sample(page_ch, 200)
```
Comment (Magicbook1108) [inline deepdoc/parser/pdf_parser.py]: This could make the results unstable and should be changed.

**Comment by asiroliu** [comment]
@tunsuy 
1. Reference the implementation code instead of duplicating it
2. Add unit tests for layout_recognizer.__is_garbage
3. Follow the existing unit test structure and place the test file in the appropriate location

File: deepdoc/parser/pdf_parser.py [hunk 2]
```diff
@@ -197,6 +198,163 @@ def _has_color(self, o):
...
+
+        for c in page_chars:
+            text = c.get("text", "")
+            fontname = c.get("fontname", "")
+            if not text or text.isspace():
+                continue
+            total_non_space += 1
+
+            if cls._has_subset_font_prefix(fontname):
+                has_subset_font = True
```
Comment (Magicbook1108) [inline deepdoc/parser/pdf_parser.py]: A single subset font can cause the entire page to be classified as font-encoding garbled.

File: deepdoc/parser/pdf_parser.py (hunk 1 above)
Comment (tunsuy) [reply deepdoc/parser/pdf_parser.py]: Fixed. Changed random.sample(page_ch, 200) to page_ch[:200] for deterministic results. Since garbled text is a page-level issue (the entire page's font encoding is broken), the first 200 characters are representative enough for detection.

File: deepdoc/parser/pdf_parser.py (hunk 2 above)
Comment (tunsuy) [reply deepdoc/parser/pdf_parser.py]: Fixed. Replaced the boolean has_subset_font flag with a subset_font_count counter and ratio check (subset_ratio = subset_font_count / total_non_space). Now the page is only considered potentially garbled when subset font characters account for ≥ 30% of all characters, avoiding false positives from pages with only a few subset font characters.

**Comment by tunsuy** [comment]
> @tunsuy
> 
> 1. Reference the implementation code instead of duplicating it
> 2. Add unit tests for layout_recognizer.__is_garbage
> 3. Follow the existing unit test structure and place the test file in the appropriate location

Okay, I've already taken care of it.

**Comment by asiroliu** [comment]
@tunsuy 
The test cases should be located in test/unit_test/deepdoc/parser/test_pdf_garbled_detection.py

**Comment by tunsuy** [comment]
> @tunsuy The test cases should be located in test/unit_test/deepdoc/parser/test_pdf_garbled_detection.py

Okay, I've already taken care of it.

**Comment by asiroliu** [comment]
Please fix CI

**Review by asiroliu — APPROVED** [review]
